Torch Data Augmentation

TimZaman opened the pull request:
Data augmentation needs little introduction I recon. It counters overfitting and makes your model generalize better, yielding better validation accuracies; or alternatively, allows you to use smaller datasets with similar performance. 

In the Zoo that's the internet, I see many implementations of different augmentations, of which few are proper and nicely portable. A part from Digits yielding a great UI; ease of use; and deep learning turn-key solution, I strongly feel we can expand to the functional side as well to make this a deep learning killer-app.

For torch, I have made an implementation _during lua preprocessing_ from frontend to backend to enable Digits to do so. In #330 there was already an attempt for augmentation, which happened on the dataset-creation side; something I am strongly against. Resizing and cropping I would consider a transformation, while I consider augmenting the data in its container an augmentation. I think therefore it's fine to resize during dataset loading (and squashing/filling/etc), but I would probably leave it at that.

Anyway, I set up a more dynamic structure to pass around these options on the torch side; instead of adding a dozen of arguments to each function, I am just adding a table.

Implements the following (screenshot): 
![image](https://cloud.githubusercontent.com/assets/7721540/15505758/258affe0-21c5-11e6-9b10-0142559ad1ef.png)

I have iterated through many augmentation types but these were the most useful. Almost done, now running elaborate tests.

# Progress

The code is already functional, though see progress below.
See code, shoot!

### Features
- [x] Make UI data transforms only visible for the Torch framework (invisible for Caffe)
- [x] ~~Implement UI option for normalization (scales the [0 255] to [0 1])~~
- [x] Data Augmentation UI
- [x] Flips (mirrors)
- [x] Quadrilateral rotations
- [x] Arbitrary rotations
- [x] Arbitrary scales
- [x] Augmenting in HSV space
- [x] Augmenting with noise (Thoughts?)
- [x] [Travis] Tests
- [x] Use Data Augmentation Template: `data_augmentation.html`

### Testing
- [x] No augmentation
- [x] Flips (mirrors)
- [x] Quadrilateral rotations
- [x] Arbitrary rotations
- [x] Arbitrary scales
- [x] Arbitrary rotations & arbitrary scales
- [x] Augmenting in HSV space
- [x] Augmenting with noise
- [x] All Augmentations & benchmark speed; identify bottlenecks
- [x] Verify models reporting a slower learning/less overfitting trade-off : more generalization.

TimZaman commented:
It this also the place to discuss the following:

- The default 'Mean Subtraction' is one 'Image'. I have never really seen any model of mine perform better than 'None', only worse. I would suggest a default to None (moreover, it's hefty on the CPU).
- `test.lua` expects square croplen images through an assert, whereas `main.lua` magically takes the minimum dimension of (croplen,x-dim,y-dim) : and might ignore the crop param altogether withouth the operator knowing about this
- The input of the image (during and after torch preprocessing) is always in the range [0 255], I guess due to the backend (lmdb/hdf5) constraints. Shouldn't we by default scale this to [0 1]? I recollect the LeNet standard network have the :mul(1/255) in there as first layer; what's the general AI network convention here? We can also make this optional. Torch's image package seems to load assume [0 1].

lukeyeager commented:
> The default 'Mean Subtraction' is one 'Image'. I have never really seen any model of mine perform better than 'None', only worse. I would suggest a default to None (moreover, it's hefty on the CPU).

Really? Interesting. I'm pretty sure LeNet won't work, at least.

And I'm also pretty sure that AlexNet and GoogLeNet both had some form of mean subtraction in their original implementations.

TimZaman commented:
> Really? Interesting. I'm pretty sure LeNet won't work, at least.
> And I'm also pretty sure that AlexNet and GoogLeNet both had some form of mean subtraction in their original implementations.

What I have come to realize, is that that Deep Learning is so complex and there are so many bells and whistles, in many ways it's not an exact science anymore, but just scientific guesses and gut feelings. If you are using mean subtraction all the time, you will probably keep on using it for forever, without really knowing why. For example: same thing goes for MaxPooling. In most cases you might as well increase the stride without any loss It's something someone thought of, seemed to work well, or it didn't, then didn't bother much about it.

Same thing goes for preprocessing to a different colorspace: RGB to HSV, or RGB with local normalization, RGB with super fancy ZCA whitening. In essence, you are never actually adding any information (obviously), you can only take information out. What you can do is stress certain parts (like you do with enhancing edge-values with local contrast or whitening); but again, deep networks can figure that out without the help of our stresses. In my experience, RGB does fine.

Lets look at what mean subtraction does for MNIST:
![image](https://cloud.githubusercontent.com/assets/7721540/15514352/008f2f9e-21e9-11e6-867c-aa1a29b2b4db.png)

My gut feeling says the mean-subtracted image probably only makes it worse (what does the mean of all these images mean anyway).

Lets investigate MNIST with default DIGITS settings:

| Type        | Max Accuracy %           | Final Accuracy %  |
| ------------- |:-------------:| -----:|
| Image      | 97.92% | 97.94% |
| Pixel      | 98.19% | 98.15% |
| None      | 98.17% | 98.16% |

The above table is in agreement with what i said earlier: why would image subtraction help? We do this originally just to put the mean at 0 because we learned from statistics that's proper. But the deep neural network can handle mostly all ranges just fine, as long as the hyper parameters are balanced to the input.

Another example then, the mean image of a Kaggle challenge i found:
![mean](https://cloud.githubusercontent.com/assets/7721540/15515011/37be82be-21ec-11e6-84c4-4ff098f7764b.jpg)
Subtracting that mean image ^ will yield really weird results.

Or the mean image of ImageNet:
![mean](https://cloud.githubusercontent.com/assets/7721540/15515095/c3e2cf98-21ec-11e6-8845-74568fdba0b0.jpg)
This gray blob won't help your ImageNet accuracy one bit ^.

In their original implementations I bet many used mean subtraction, but that does not mean it yields the best results, which i think is more important.

See if you can reproduce the results from the above table. Why did you say it won't work for LeNet?

gheinrich commented:
This can be debated at length but a reference paper on the subject is [Efficient Backprop](http://yann.lecun.com/exdb/publis/pdf/lecun-98b.pdf). Normalizing inputs and initializing weights carefully aims at sticking to the range where transfer functions are nicely non-linear and have non-zero gradients. You can always get around this after some learning (the network will learn how much bias to apply to the receptive fields) but sometimes it is almost impossible to learn efficiently. If the input to e.g. a sigmoid is very large then the gradient is almost zero and it is very difficult to learn.

In the example from the regression tutorial I've seen cases where the network diverges (loss->NaN) without mean subtraction. This has never happened with mean subtraction. The learning curve is smoother with mean image subtraction:
![screenshot from 2016-05-24 21 50 35](https://cloud.githubusercontent.com/assets/3889770/15517717/990990e2-21f9-11e6-8cf1-c46b5f9f749a.png)
Compare with no mean subtraction:
![screenshot from 2016-05-24 21 51 44](https://cloud.githubusercontent.com/assets/3889770/15517747/bc8c8ae2-21f9-11e6-9598-c0234fa2f9c8.png)

The common practice is to normalize inputs so I guess that's what we need to do by default in DIGITS. On that subject you're making a good point about the pixel range:
> The input of the image (during and after torch preprocessing) is always in the range [0 255]

I agree it would make more sense to be in the [0 1] range. We could very well do that in Torch but the thing is this isn't what Caffe is doing. Since DIGITS originally only supported Caffe I opted for the [0 255] range to get as close a user experience as possible. I fear this will make us suffer when we want to support 16-bit images!

Thanks a lot for the PR this is great!

lukeyeager commented:
> See if you can reproduce the results from the above table. Why did you say it won't work for LeNet?

Oh you're right! I was thinking of something else, related to something else you brought up:

> The input of the image (during and after torch preprocessing) is always in the range [0 255], I guess due to the backend (lmdb/hdf5) constraints. Shouldn't we by default scale this to [0 1]? I recollect the LeNet standard network have the :mul(1/255) in there as first layer; what's the general AI network convention here? We can also make normalization optional?

AlexNet and GoogLeNet both work for [0-1] and [0-255] data, but LeNet only works on [0-1] data.

So nevermind about the mean subtraction - my bad!

gheinrich commented:
> Make UI data transforms only visible for the Torch framework (invisible for Caffe)

For this you could have a look at the [data shuffle capability](https://github.com/NVIDIA/DIGITS/blob/v3.3.0/digits/frameworks/framework.py#L22) and how it's used in the [templates](https://github.com/NVIDIA/DIGITS/blob/v3.3.0/digits/templates/models/images/classification/new.html#L401) to selectively enable the corresponding form field.

gheinrich commented on digits/templates/models/images/classification/new.html:
```diff
@@ -356,16 +356,104 @@ <h4 style="display:inline-block;">Python Layers</h4>
         <div class="col-sm-4">
             <div class="well">
                 <h4>Data Transformations</h4>
+                <div class="form-group{{mark_errors([form.use_mean])}}">
```
I feel we haven't done enough of this but you could perhaps move this section to a `data_augmentation.html' template and include it here and in `digits/templates/models/images/generic/new.html`

TimZaman commented:
Come to think of it, indeed I think LeNet was meant for bitonal images, so
probably no default mean subtraction at all. Moreover, you don't need mean
subtraction from the dataset to yield a range that complies with the paper
you are refering to, we do not need to discuss that at length at all,
because you are right. But as just shown with MNIST, mean-subtraction by
the pixel-for-pixel mean of all images the dataset will never yield better
results; especially when the dataset does not contain hundreds of thousands
of images. Therefore i coined not using 'Image' as mean subtraction by
default.

On Tue, May 24, 2016 at 10:46 PM, Luke Yeager <notifications@github.com>
wrote:

> See if you can reproduce the results from the above table. Why did you say
> it won't work for LeNet?
>
> Oh you're right! I was thinking of something else, related to something
> else you brought up:
>
> The input of the image (during and after torch preprocessing) is always in
> the range [0 255], I guess due to the backend (lmdb/hdf5) constraints.
> Shouldn't we by default scale this to [0 1]? I recollect the LeNet standard
> network have the :mul(1/255) in there as first layer; what's the general AI
> network convention here? We can also make normalization optional?
>
> AlexNet and GoogLeNet both work for [0-1] and [0-255] data, but LeNet only
> works on [0-1] data.
>
> So nevermind about the mean subtraction - my bad!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/NVIDIA/DIGITS/pull/777#issuecomment-221385082>
>

TimZaman commented:
>For this you could have a look at the data shuffle capability and how it's used in the templates to selectively enable the corresponding form field.
Thanks, that helped. I made it fairly dynamic now, and made a neat template. I took the liberty of putting the template in `digits/templates/models/` because it can be used by a generic class and classification one, hope that's allright.
Realizing that, in efforts to reduce redundancy we can merge the identical files:
`generic/large_graph.html` with `classification/large_graph.html`
`generic/custom_network_explanation.html` with `classification/custom_network_explanation.html`
and maybe put those in a more general folder.

pansk commented:
I like this feature, that's basically more than what I've done in my matlab preprocessing step (except the multiscale extraction - I also extract regions from the same image downsampled 2x, 4x, etc)

About mean subtraction, in my experience with autoencoders it did help (at least in the first tests I took), so I ended up using it by default. I'll run a test on one of the latest models right now, and see what happens.

TimZaman commented:
> I like this feature
Thanks. Do you have any other augmentation steps that work well? For example, I have not included 'blurring' because it seemed to be relatively ineffective. How is multiscale extraction working for you?

> About mean subtraction, in my experience with autoencoders it did help (at least in the first tests I took), so I ended up using it by default.
Sure it helps for the reasons accurately described above. But using 'Image' subtraction as opposed to 'Pixel' subtraction will probably not make an improvement I think.

gheinrich commented:
> using 'Image' subtraction as opposed to 'Pixel' subtraction will probably not make an improvement I think

I agree: image subtraction probably only helps for MNIST where DIGITS are nicely centred in the image but for realistic datasets pixel subtraction might make more sense. Besides, image subtraction is painful to work with for networks that accept various inputs sizes like FCNs.

pansk commented:
@TimZaman sure, I'd gladly test!
Another augmentation I would like is a filtering pass (lowpass/highpass), but I think it's very application-specific, and it's probably not worth to generalise this step.

What about adding noise (with different scales)? E.g. augment a set adding images with 0%, 1%, 2%, 3%, 5%, 10% noise (list of noise % specified by the user).

> Sure it helps for the reasons accurately described above. But using 'Image' subtraction as opposed to 'Pixel' subtraction will probably not make an improvement I think.

I ran a full training on a previous network, removing the mean subtraction: the resulting quality is unchanged. Thank you for pointing this out! 👍

TimZaman commented:
High pass and low pass filtering.. Need to think about that one, how it
would help generalize. How do you use it now, DFT, or something simple?
Adding noise or blurring (blurring would already be low pass filtering i
guess) is straightforward and we can put it in, but often time i have seen
people reporting that it's almost the same as adding dropout: you're not
changing the image significantly enough; compare to a horizontal flip. A
horizontal flip for something like natural images almost doubles your
dataset because the image looks quite different. Whereas perturbing pixels
with noise is like dropping out some pixels.
Having said that, I might as well try. What are your experienced there?

On Thursday, 26 May 2016, Marco Foco <notifications@github.com> wrote:

> @TimZaman <https://github.com/TimZaman> sure, I'd gladly test!
> Another augmentation I would like is a filtering pass (lowpass/highpass),
> but I think it's very application-specific, and it's probably not worth to
> generalise this step.
>
> What about adding noise (with different scales)? E.g. augment a set adding
> images with 0%, 1%, 2%, 3%, 5%, 10% noise (list of noise % specified by the
> user).
>
> Sure it helps for the reasons accurately described above. But using
> 'Image' subtraction as opposed to 'Pixel' subtraction will probably not
> make an improvement I think.
>
> I ran a full training on a previous network, removing the mean
> subtraction: the resulting quality is unchanged. Thank you for pointing
> this out! 👍
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/NVIDIA/DIGITS/pull/777#issuecomment-221988941>
>

TimZaman commented:
Stumbled upon some bugs in torch/image while testing scale and rotation. https://github.com/torch/image/pull/169
Now running CIFAR10 benchmarks for many augmentation types.

pansk commented:
I have tried both adding noise to original images (huge amount of training/test data is produced) and using dropout in place of noise.
I think adding a known amount of noise is more straightforward, since you can obtain the performance values for different noise scales directly from your test results.

TimZaman commented:
> I think adding a known amount of noise is more straightforward, since you can obtain the performance values for different noise scales directly from your test results.
Okay I'll see if I can add that. What kind of noise do you suggest, and what kind of underlying distribution? Matrix of a normal distribution then multiplying? Each channel seperate; or same for each? We're gonna need to make a ton of assumptions :).

gheinrich commented on digits/model/images/forms.py:
```diff
@@ -34,3 +30,81 @@ class ImageModelForm(ModelForm):
+                ('rot90', '0, 90 or 270 degrees'),
+                ('rot180', '0 or 180 degrees'),
+                ('rotall', '0, 90, 180 or 270 degrees.'),
+                ],
```
are `rot180` and `rotall` useful? Rotation by 180 degrees is the same as vertical flipping.

TimZaman replied:
Almost. A vertical flip + horizontal flip = 180 rotation.

gheinrich commented on digits/model/images/forms.py:
```diff
@@ -34,3 +30,81 @@ class ImageModelForm(ModelForm):
+            default = False,
+            tooltip = "You can augment your dataset by performing rotation on images.",
+            validators=[ 
+                ]
```
that is a just a special case of uniform-random rotation of +/- 180 degrees, right?

TimZaman replied:
It's an arbitrary rotation in the sense that it's a rotation between [-rot rot] degrees, distributed uniformly. How should I call this?

gheinrich replied:
sorry I didn't understand the field correctly. I think the field could be removed if we have the next one default to 0.

gheinrich commented on digits/model/images/forms.py:
```diff
@@ -34,3 +30,81 @@ class ImageModelForm(ModelForm):
+    aug_rot = utils.forms.IntegerField('Rotation (+- deg)',
+            default=45,
+            validators=[ 
+                validators.NumberRange(min=0, max=360)
```
Why not make it `max=180` since the rotation is +/- the specified angle?

gheinrich commented on digits/model/images/forms.py:
```diff
@@ -34,3 +30,81 @@ class ImageModelForm(ModelForm):
+            tooltip = "The uniform-random rotation angle that will be performed during batch preprocessing."
+            )
+
+    aug_scale_use = utils.forms.BooleanField('Random Scaling',
```
can we remove this field and have the rescale std default to 0?

TimZaman replied:
Yes we can. Do you want me to? I was following the redundant convention of the crop and croplen (in main.lua, see below), but extended it in this case to the frontend.
```
-k,--crop (default no) If this option is 'yes', all the images are randomly cropped into square image. And croplength is provided as --croplen parameter
-l,--croplen (default 0) crop length. This is required parameter when crop option is provided
```

gheinrich replied:
Cropping is different: there is no value of the crop length that yields no transformation.

I'm in favour of removing the boolean field here and rely on just the `aug_scale` field instead but I'm open to discussing it :-)

TimZaman replied:
I am inclined to agree, unless we think of options to expand the scale augmentation type. For example: the statistical distribution type; only zooming out/only zooming im; etc. But I feel i get too technical there and clutter with insignificant geek/nerd options.

gheinrich commented on digits/model/images/forms.py:
```diff
@@ -34,3 +30,81 @@ class ImageModelForm(ModelForm):
+            tooltip = "Retaining image size, the image is rescaled with a +-stddev of this parameter."
+            )
+
+    aug_hsv_use = utils.forms.BooleanField('HSV Shifting',
```
maybe this one can also be removed if the below fields default to 0?

gheinrich commented on tools/torch/data.lua:
```diff
@@ -41,33 +41,190 @@ local function all_keys(cursor_,key_,op_)
+-- @param meanTensor (tensor): mean image that will be used for mean subtraction if supplied
+-- @param augOpt (table): structure containing several augmentation options.
+local PreProcess = function(im, train, meanTensor, augOpt)
+--image.savePNG( '/home/brainstorm/tmp/' .. os.date("%H%I%M%S_before_mean") ..'.png', torch.div(im,255))
```
this line and line 174 could be removed in the final version of this patch :-)

TimZaman replied:
Sir Yes Sir

TimZaman commented:
Initial results. Trained on CIFAR10 with a great VGG network with the overfit we love to see. (https://github.com/szagoruyko/cifar.torch/blob/master/models/vgg_bn_drop.lua)

The results reveal the augmentation is working really nicely. Training speed per epoch does not seem to be impacted.

# Scale
![scale](https://cloud.githubusercontent.com/assets/7721540/15613142/3e24f294-2432-11e6-8979-85823a26657a.gif)

# Rotation
![arbrot](https://cloud.githubusercontent.com/assets/7721540/15613146/4553e232-2432-11e6-8798-9d9309f7c37c.gif)

# Flipping
![hflip](https://cloud.githubusercontent.com/assets/7721540/15613144/43f034cc-2432-11e6-8b6e-4d03eaffe63e.gif)

TimZaman commented:
# All Augmentations
HFlip, rotation 5, scale stddev 0.05, hsv (0.01,0.02,0.04) 
**yields a full 3 percent increase**

![all](https://cloud.githubusercontent.com/assets/7721540/15646584/b310abba-265e-11e6-9204-c971d850903f.gif)

gheinrich commented:
That is totally awesome! That is a truly great feature! Can't wait to have that merged.

Do you think you can add tests for this (not necessarily to test that augmentation reduces overfit but at least to exercise the new code in the automatic tests). Do you need help there?

TimZaman commented:
> Do you think you can add tests for this (...). Do you need help there?

Yes I need some help indeed; a few pointers would be great. Which tests do you suggest and where do you suggest I put them? I have not looked into how this is done in this project at all.

pansk commented:
I'd start with AWGN (Additive White Gaussian Noise) because it's easier to generate (just adding a 0-mean normal-distributed random variable will do), and you just need one parameter (SNR or standard deviation) to describe it (if you assume the signal's power at 0dB).
It is also a good approximation for thermal noise.

TimZaman commented:
> I'd start with AWGN
Seems the most straightforward. So just generate a #channel WGN tensor, then add it to the image tensor, clamp/saturate between [0 255], i guess.

gheinrich commented:
You could look at #797 which adds a couple of tests for generic image models trained on generic datasets. To run tests on your machine you will need the modules from https://github.com/NVIDIA/DIGITS/blob/v3.3.0/requirements_test.txt.

To run tests from a single test class you may do for example (from the root of DIGITS):
```
./digits-test -v -s digits/model/images/generic/test_views.py:TestCaffeCreated
```

For classifications models, tests are located in https://github.com/NVIDIA/DIGITS/blob/v3.3.0/digits/model/images/classification/test_views.py. For a simple example, check how we added test cases to check multistep-learning-rate policy: 
https://github.com/NVIDIA/DIGITS/blob/v3.3.0/digits/model/images/classification/test_views.py#L1004-L1007

Those test classes rely on class inheritance quite heavily and might require some time to get used to :-( let us know if you need more info. Thanks!

TimZaman commented:
## Resnet
ResNet's can overfit like crazy, meaning there's ton to be gained from augmentation. And yep, from 83% to over 91% with added augmentation.
Note in the below graphs the epochs start at 10 as opposed to 0 to show the loss-value degressions in a better scale.

### No Augmentation
![image](https://cloud.githubusercontent.com/assets/7721540/15752664/0f495e74-28e7-11e6-8f00-1f6676565b2f.png)

### Augmentation
Augmentation values used: horizontal flip, rotation 10, scale 0.07, hsv (0.02, 0.03, 0.06)
![image](https://cloud.githubusercontent.com/assets/7721540/15752653/02c93d7c-28e7-11e6-961a-6c1d898d3e6a.png)

TimZaman commented:
@pansk 
![image](https://cloud.githubusercontent.com/assets/7721540/15799927/5f9df38a-2a6d-11e6-84c0-ad72723eabc5.png)
I'd like to take you up on your offer of checking the augmentation types! Your AWGN is in there too now.
@gheinrich 
I have simplified the view as you requested, made much more sense indeed:
![image](https://cloud.githubusercontent.com/assets/7721540/15799933/8272cd40-2a6d-11e6-912a-8528594db119.png)
All that's left is the tests now.

TimZaman commented:
> Those test classes rely on class inheritance quite heavily and might require some time to get used to :-( 

Haha, I can see what you meant! Seems to work well. I just added a test to see if the augmentation parameters all work. It worked very well because when I had it implemented for the generic one, it failed because HSV augmentation needs 3 channels. But i removed the checks from the generic one altogether because I felt it was too redundant with the classification one.

I _think_ this is now done and ready for review.

TimZaman commented:
And last but not least (@pansk), the Noise graph. You can clearly see I have overestimated the value the stddev should be. Note only a very small amount of noise helps, and it does not help very much, but it does a bit when the value is small enough. The problem I do have with this approach though is that all images are noisy. My gut feeling tells me it's good to noise it, but even have a second parameter that randomizes the standard deviation parameter _itself_ to constrain the overall noisyness in some images.

### Noise
Values used: 0, 0.01, 0.03, 0.06, 0.09. I should have used more values around the 0.01 mark but ok.
![noise](https://cloud.githubusercontent.com/assets/7721540/15804399/047ba458-2b07-11e6-82e1-10d059da7286.gif)

pansk commented:
I switched to this branch for testing, and suddenly found out some problems:

- with a db of ~100 2048x1536x3 images, it goes immediately out of memory (I suppose it's the GPU, in my case a 980Ti). It's not an issue of this PR, but it's a possible optimization area.

- If I enable HSV augmentation, I receive this error:

2016-06-06 16:29:22 [20160606-162921-566a] [INFO ] Train Torch Model task started.
2016-06-06 16:29:29 [20160606-162921-566a] [DEBUG] Network validation loss #0.0: 19480.0644531
2016-06-06 16:29:29 [20160606-162921-566a] [ERROR] Train Torch Model: /torch/install/share/lua/5.1/threads/threads.lua:183: [thread 1 callback] inconsistent tensor size at /torch/pkg/torch/lib/TH/generic/THTensorMath.c:563
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: stack traceback:
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: [C]: at 0x7f47607efae0
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: [C]: in function 'add'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /digits/tools/torch/data.lua:54: in function 'augmentHSV'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /digits/tools/torch/data.lua:184: in function 'PreProcess'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /digits/tools/torch/data.lua:658: in function 'nextBatch'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /digits/tools/torch/data.lua:783: in function </digits/tools/torch/data.lua:777>
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: [C]: in function 'xpcall'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /torch/install/share/lua/5.1/threads/threads.lua:234: in function 'callback'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /torch/install/share/lua/5.1/threads/queue.lua:65: in function </torch/install/share/lua/5.1/threads/queue.lua:41>
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: [C]: in function 'pcall'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /torch/install/share/lua/5.1/threads/queue.lua:40: in function 'dojob'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: [string "  local Queue = require 'threads.queue'..."]:15: in main chunk
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: DIGITS Lua Error
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: stack traceback:
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: [C]: in function 'error'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /torch/install/share/lua/5.1/threads/threads.lua:183: in function 'dojob'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /digits/tools/torch/data.lua:806: in function 'waitNext'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /digits/tools/torch/main.lua:749: in function 'Train'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /digits/tools/torch/main.lua:842: in main chunk
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: [C]: in function 'dofile'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /digits/tools/torch/wrapper.lua:25: in function </digits/tools/torch/wrapper.lua:25>
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: [C]: in function 'xpcall'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /digits/tools/torch/wrapper.lua:25: in main chunk
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: [C]: in function 'dofile'
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: /torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model unrecognized output: [C]: at 0x00406670
2016-06-06 16:29:30 [20160606-162921-566a] [ERROR] Train Torch Model task failed with error code 1

Removing HSV augmentation the process starts correctly.

- How about adding a "number of samples" option, to extract more than one patch from the same image?

Also, I couldn't one thing: if my region lies outside the original image (because of rotation or scale jittering) what's going to happen? Is the border going to be replicated, or filled with black? Or is the sample is just discarded?

(I'm doing the first training now, I'll come back with my results later today or tomorrow)

TimZaman commented:
Yeah 2048x1536x3 is HUGE if you have a big model as well :). So indeed your
first memory error isnt related to this PR, also because the augmentations
are happening on the cpu.
I will look into the hsv error, it has worked for me, might be batch-size
related or something.

Please elaborate on the 'number of samples' option and when this comes in
handy.

Borders are being replicated. During a zoom in or zoom out the source image
is randomly translated throughout the canvas.
I almost feel i should make a help HTML-explanation for this.



On Monday, 6 June 2016, Marco Foco <notifications@github.com> wrote:

> I switched to this branch for testing, and suddenly found out some
> problems:
>
>    -
>
>    with a db of ~100 2048x1536x3 images, it goes immediately out of
>    memory (I suppose it's the GPU, in my case a 980Ti). It's not an issue of
>    this PR, but it's a possible optimization area.
>    -
>
>    If I enable HSV augmentation, I receive this error:
>
> 2016-06-06 16:29:22 [20160606-162921-566a] [INFO ] Train Torch Model task
> started.
> 2016-06-06 16:29:29 [20160606-162921-566a] [DEBUG] Network validation loss
> #0.0: 19480.0644531
> 2016-06-06 16:29:29 [20160606-162921-566a] [ERROR] Train Torch Model:
> /torch/install/share/lua/5.1/threads/threads.lua:183: [thread 1 callback]
> inconsistent tensor size at
> /torch/pkg/torch/lib/TH/generic/THTensorMath.c:563
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: stack traceback:
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: [C]: at 0x7f47607efae0
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: [C]: in function 'add'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: /digits/tools/torch/data.lua:54: in function
> 'augmentHSV'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: /digits/tools/torch/data.lua:184: in function
> 'PreProcess'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: /digits/tools/torch/data.lua:658: in function
> 'nextBatch'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: /digits/tools/torch/data.lua:783: in function
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: [C]: in function 'xpcall'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: /torch/install/share/lua/5.1/threads/threads.lua:234:
> in function 'callback'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: /torch/install/share/lua/5.1/threads/queue.lua:65: in
> function
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: [C]: in function 'pcall'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: /torch/install/share/lua/5.1/threads/queue.lua:40: in
> function 'dojob'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: [string " local Queue = require
> 'threads.queue'..."]:15: in main chunk
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: DIGITS Lua Error
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: stack traceback:
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: [C]: in function 'error'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: /torch/install/share/lua/5.1/threads/threads.lua:183:
> in function 'dojob'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: /digits/tools/torch/data.lua:806: in function
> 'waitNext'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: /digits/tools/torch/main.lua:749: in function 'Train'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: /digits/tools/torch/main.lua:842: in main chunk
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: [C]: in function 'dofile'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: /digits/tools/torch/wrapper.lua:25: in function
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: [C]: in function 'xpcall'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: /digits/tools/torch/wrapper.lua:25: in main chunk
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: [C]: in function 'dofile'
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output:
> /torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
> 2016-06-06 16:29:29 [20160606-162921-566a] [WARNING] Train Torch Model
> unrecognized output: [C]: at 0x00406670
> 2016-06-06 16:29:30 [20160606-162921-566a] [ERROR] Train Torch Model task
> failed with error code 1
>
> Removing HSV augmentation the process starts correctly.
>
>    - How about adding a "number of samples" option, to extract more than
>    one patch from the same image?
>
> Also, I couldn't one thing: if my region lies outside the original image
> (because of rotation or scale jittering) what's going to happen? Is the
> border going to be replicated, or filled with black? Or is the sample is
> just discarded?
>
> (I'm doing the first training now, I'll come back with my results later
> today or tomorrow)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIA/DIGITS/pull/777#issuecomment-223979994>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AHXSRDOlLdyqN9mk5YAaL89MIgwwv0k8ks5qJDIGgaJpZM4Ilgr1>
> .
>

pansk commented:
Oh, I see, apparently I forgot to mention that my crop size is 128 :)
The model is not that big after all :)

TimZaman commented:
@pansk 
Cropping happens at the very end.
Off-topic: I am in favor of using rescaling over cropping. A rescale augmentation is more methodologically correct in my opinion, mostly due to the fact that with cropping you generally have a problem of 'how do I validate my source images', which in Digits-torch is a center-crop.
If you stop using cropping, and use rescaling, you can still (when zoomed in or out) wiggle your image around the canvas: just as cropping (think about it..). Your mean 'rescale' will be 1:1, just like your validation pass image will be used entirely and 1:1.
But, for historic reasons everyone just uses cropping.

gheinrich commented on digits/templates/models/data_augmentation.html:
```diff
@@ -0,0 +1,67 @@
+{# Copyright (c) 2014-2016, NVIDIA CORPORATION.  All rights reserved. #}
```
minor comment: you can make this `2016` only since this is a new file

gheinrich commented on tools/torch/data.lua:
```diff
@@ -41,33 +41,205 @@ local function all_keys(cursor_,key_,op_)
+    if augHSV.H >0.001 then
+        -- We do not need to account for overflow because every number wraps around (1.1=2.1=3.1,etc)
+        local rng_h = 1 + torch.normal(0, augHSV.H) -- We add a round value 1 to prevent an underflow bug (<0 becomes black bug)
+        im_hsv[1] = torch.add(im_hsv[1], torch.FloatTensor(1,im_hsv:size(2),im_hsv:size(2)):fill(rng_h))
```
```
torch.FloatTensor(1,im_hsv:size(2),im_hsv:size(2))
```

shouldn't this be:
```
torch.FloatTensor(1,im_hsv:size(2),im_hsv:size(3))
```

TimZaman replied:
Indeed it should.

gheinrich replied:
I think this can be simplified a bit: you can add a scalar to a tensor.
```lua
th> x=torch.Tensor(2,3):fill(0)
                                                                      [0.0001s]	
th> x
 0  0  0
 0  0  0
[torch.DoubleTensor of size 2x3]

                                                                      [0.0002s]	
th> x+torch.normal(0,0.1)
-0.1900 -0.1900 -0.1900
-0.1900 -0.1900 -0.1900
[torch.DoubleTensor of size 2x3]

                                                                      [0.0002s]	
th> x+torch.normal(0,0.1)
0.01 *
 -1.6087 -1.6087 -1.6087
 -1.6087 -1.6087 -1.6087
[torch.DoubleTensor of size 2x3]
```

TimZaman replied:
Good! I think that's more efficient. I'll have to add :float() to the scalar though; the color transformation cries about double's.

gheinrich commented on tools/torch/data.lua:
```diff
@@ -41,33 +41,205 @@ local function all_keys(cursor_,key_,op_)
+local augmentHSV = function(im_rgb, augHSV)
+    -- Fair augHSV standard deviation values are {H=0.02,S=0.04,V=0.08}
+    local im_hsv = image.rgb2hsv(im_rgb)
+    if augHSV.H >0.001 then
```
why not `augHSV.H > 0`? I'm sure there is a reason, I'm just asking :-)

TimZaman replied:
Hehe there's a reason. Because any value below 0.001 will do so little it will make no sense at all. So it it might save some computation time here and there, especially with very little H-augmentation and a standard deviation like 0.01. For readability sake I could also take it out.

gheinrich replied:
Ah OK. Isn't it the responsibility of the user to choose a value that makes sense? I'm worried the magic value here could be confusing.

TimZaman replied:
You are right. Will change.

gheinrich commented on digits/model/images/classification/test_views.py:
```diff
@@ -825,6 +854,17 @@ class BaseTestCreatedTall(BaseTestCreated):
 class BaseTestCreatedCropInForm(BaseTestCreated):
     CROP_SIZE = 8
 
+class BaseTestCreatedDataAug(BaseTestCreated):
```
if you inherit from `BaseTestCreatedTall` you'll get to test this on non-square images. Wouldn't that expose the issue on line 54 of data.lua?

TimZaman replied:
Indeed! Which ones would you suggest? And the square, and the tall and the wide or just one? Note also it has to be 3 channels to be able to do HSV augmentation.

gheinrich replied:
I don't know... it feels like one class is enough (e.g. `Tall`) but you never know...

gheinrich commented on tools/torch/data.lua:
```diff
@@ -41,33 +41,205 @@ local function all_keys(cursor_,key_,op_)
+-- @param augScale (float): the standard deviation of the extra scaling factor
+local warp = function(im, augRot, augScale)
+    -- A nice function of scale is 0.05 (stddev of scale change), 
+    -- and a nice value for ration is a few degrees or more if your dataset allows for it
```
"a nive value for **rotation**"?

TimZaman replied:
"a **nice** value for rotation"?

gheinrich replied:
Arf! Yes :-)

gheinrich commented on tools/torch/data.lua:
```diff
@@ -41,33 +41,205 @@ local function all_keys(cursor_,key_,op_)
+    end
+
+    -- Angle of rotation
+    local rot_angle = (torch.rand(1)[1]-0.5)* 2 * augRot -- (degrees) uniform distribution [-augRot : augRot)
```
wondering why you're not doing `rot_angle = torch.uniform(-augRot,augRot)`?

TimZaman replied:
That looks much cleaner indeed.

gheinrich commented on tools/torch/data.lua:
```diff
@@ -41,33 +41,205 @@ local function all_keys(cursor_,key_,op_)
+        flow:add(flow_rot)
+    end
+
+    return image.warp(im, flow, 'bilinear', false)
```
did you choose to use `image.warp` because it does scaling and rotation in a single pass so it's more efficient that using `image.scale` and `image.rotate`?

TimZaman replied:
That, and the fact that scale and rotate do not pad the images with a nice looking replication. And when the two operations are independent, the first operation might cut off a part in the image that might have been revealed due to the next operation (for example, if i first rotate, then zoom out, i lose the corners of my original image while they could have been retained).

gheinrich commented on tools/torch/data.lua:
```diff
@@ -489,34 +674,33 @@ function DBSource:totalRecords ()
 
 -- Derived class method close (LMDB flavour)
-function DBSource:lmdb_close (batchsize)
+function DBSource:lmdb_close (batchSize)
```
not related to this patch but that makes me wonder why we have `batchSize` here... probably a left-over from a previous patch

TimZaman replied:
Shall I remove it while I'm at it? Note it's also in the other db closure `DBSource:hdf5_close (batchSize)`

gheinrich replied:
If you don't mind, yes. Thanks.

pansk commented:
I'm sorry for the delay, I'm verifying, because it doesn't seem to converge well, and I want to be sure it's not something related to my specific model.

TimZaman commented:
Strange. What kind of model are you using, and dataset? I verified this to
work well on a few different datasets, although the hsv augmentation could
be slightly improved. Especially the horizontal flipping is very
straightforward and should help. rotation of a few degrees helps on most
datasets, and scaling too.

On Wed, Jun 8, 2016 at 7:49 PM, Marco Foco <notifications@github.com> wrote:

> I'm sorry for the delay, I'm verifying, because it doesn't seem to
> converge well, and I want to be sure it's not something related to my
> specific model.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIA/DIGITS/pull/777#issuecomment-224672799>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AHXSRN2xugocaVv-GxgrUFEndDHbXWMHks5qJwCigaJpZM4Ilgr1>
> .
>

pansk commented:
I'm pretty sure it's my code, that's why I'm going to try with a simpler autoencoder model, but I've to build a specific dataset for that (the MNIST is grayscale only).

pansk commented:
Just for reference, I was going to post this comment few days ago:

I'm training an autoencoder with some custom datasets with uniform images (like wood, stone walls, grass, and so on), so I don't really need to validate my input, but border effects might be annoying for me. On the other hand, I'd like to provide just a bunch of pictures (eventually, of different sizes) and let DIGITS generate its own set of images extracting random regions, rotating, scaling, adding noise, and so on. That's why I'd like to extract more than one region per source image.

My parameters were:

![screenshot from 2016-06-07 11 23 09](https://cloud.githubusercontent.com/assets/1263535/15852733/a0bc1cac-2ca2-11e6-948c-fd70f2aa6ed6.png)

For the general case, I agree for cropping vs. scaling (actually, scaling gives you more robustness).

Anyway, are you sure digits-torch just uses center-cropping? If so, the baloon help for crop parameter is a bit misleading: "If specified, during a training a random square crop will be taken from the input image before using as input for the network."

gheinrich commented:
> are you sure digits-torch just uses center-cropping?

@TimZaman probably meant to say that we are doing centre cropping during validation. During training, we are doing random cropping.

TimZaman commented:
> > are you sure digits-torch just uses center-cropping?

>  @TimZaman probably meant to say that we are doing centre cropping during validation. During training, we are doing random cropping.

Correct, I said verbatim

> (...) with cropping you generally have a problem of **'how do I validate my source images'**, which in Digits-torch is a **center-crop**.

pansk commented:
Sorry Tim, I didn't notice you were referring to validation-only.
Do any of you know the reason behind this choice?

TimZaman commented:
It's the most valid choice. If you want to be more fancy you can do a few crops and take the average (or the max before normalizing) from those. There's really no 'correct' way of validating when you are cropping, since you have to cut parts of your source image off during validation because your model is smaller than your actual images. But in practice, performance wise - it doesn't matter much.

TimZaman commented:
With the latest revision, and If Travis agrees with me, this PR is done I think.

gheinrich commented:
This all looks good to me, thanks for the awesome PR! Can you squash your commits (possibly rebase too)?

TimZaman commented:
Squashed & rebased.

gheinrich commented:
Splendid!

philipperemy commented:
Any updates since then?

lukeyeager commented:
This looks good to me!

@gheinrich please merge unless you have more concerns.

TimZaman commented:
The 'data hook' idea has also grown on me. I think it's a great (and pretty
straightforward) feature. But I do like the UI this PR brings for obvious
reasons. It's too bad Caffe doesnt have great augmentation layers in its
master, although there are some pro forks.
Also, there is a proper PR in Caffe for a confusion matrix layer. That
would mean Caffe could show it during training if it's added, as with
torch, it can be easily captured with a regex. Mkay I digress.

On Tuesday, 26 July 2016, Luke Yeager <notifications@github.com> wrote:

> This looks good to me!
>
> @gheinrich <https://github.com/gheinrich> please merge unless you have
> more concerns.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIA/DIGITS/pull/777#issuecomment-235113978>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHXSRBiVbfbgfO6l8fF9mXTvyjUvnFAXks5qZUIqgaJpZM4Ilgr1>
> .
>
